# Nanodegree Engenheiro de Machine Learning
## Aprendizagem Supervisionada
## Projeto 2: Construindo um Sistema de Intervenção para Estudantes

Bem-vindo ao segundo projeto do Nanodegree de Machine Learning! Neste Notebook, alguns templates de código já foram fornecidos, e será o seu trabalho implementar funcionalidades necessárias para completar este projeto com êxito. Seções que começam com **'Implementação'** no cabeçalho indicam que o bloco de código que se segue precisará de funcionalidades adicionais que você deve fornecer. Instruções serão providenciadas para cada seção e as especificações para cada implementação estarão marcadas no bloco de código com o comando `'TODO'`. Tenha certeza de ler atentamente todas as instruções!

Além do código implementado, haverá questões relacionadas ao projeto e à implementação que você deve responder. Cada seção em que você tem que responder uma questão será antecedida de um cabeçalho **'Questão X'**. Leia atentamente cada questão e escreva respostas completas nas caixas de texto subsequentes que começam com **'Resposta: '**. O projeto enviado será avaliado baseado nas respostas para cada questão e a implementação que você forneceu.  

>**Nota:** Células de código e Markdown podem ser executadas utilizando o atalho de teclado **Shift + Enter**. Além disso, as células Markdown podem ser editadas, um clique duplo na célula entra no modo de edição.

### Questão 1 - Classificação versus Regressão
*Seu objetivo neste projeto é identificar estudantes que possam precisar de intervenção antecipada antes de serem reprovados. Que tipo de problema de aprendizagem supervisionada é esse: classificação ou regressão? Por quê?*

**Resposta: Classificação, pois a váriavel resposta não é contínua.**

## Observando os Dados
Execute a célula de código abaixo para carregar as bibliotecas de Python necessárias e os dados sobre os estudantes. Note que a última coluna desse conjunto de dados, `'passed'`, será nosso rótulo alvo (se o aluno foi ou não aprovado). As outras colunas são atributos sobre cada aluno.

In [48]:
# Importar bibliotecas
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Ler os dados dos estudantes
student_data = pd.read_csv("student-data.csv")
print ("Os dados dos estudantes foram lidos com êxito!")

Os dados dos estudantes foram lidos com êxito!


In [49]:
student_data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   ...   internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0  ...         no       no       4         3      4    1    1      3        6   
1  ...        yes       no       5         3      3    1    1      3        4   
2  ...        yes       no       4         3      2    2    3      3       10   
3  ...        yes      yes       3         2      2    1    1      5        2   
4  ...         no       no       4         3      2    1    2      5        4   

  passed  
0     no  
1     no  
2    yes  
3    yes  
4    yes  

[5 rows x 31 columns]

### Implementação: Observando os Dados
Vamos começar observando o conjunto de dados para determinar quantos são os estudantes sobre os quais temos informações e entender a taxa de graduação entre esses estudantes. Na célula de código abaixo, você vai precisar calcular o seguinte:
- O número total de estudantes, `n_students`.
- O número total de atributos para cada estudante, `n_features`.
- O número de estudantes aprovados, `n_passed`.
- O número de estudantes reprovados, `n_failed`.
- A taxa de graduação da classe, `grad_rate`, em porcentagem (%).


In [50]:
# TODO: Calcule o número de estudante
n_students = student_data.shape[0]

# TODO: Calcule o número de atributos
n_features = student_data.shape[1] - 1

# TODO: Calcule o número de alunos aprovados
n_passed = student_data[student_data['passed'] == 'yes'].shape[0]

# TODO: Calcule o número de alunos reprovados
n_failed = student_data[student_data['passed'] == 'no'].shape[0]

# TODO: Calcule a taxa de graduação
grad_rate = float(n_passed/n_students)*100

# Imprima os resultados
print ("Número total de estudantes: {}".format(n_students))
print ("Número de atributos: {}".format(n_features))
print ("Número de estudantes aprovados: {}".format(n_passed))
print ("Número de estudantes reprovados: {}".format(n_failed))
print ("Taxa de graduação: {:.2f}%".format(grad_rate))

Número total de estudantes: 395
Número de atributos: 30
Número de estudantes aprovados: 265
Número de estudantes reprovados: 130
Taxa de graduação: 67.09%


## Preparando os Dados
Nesta seção, vamos preparara os dados para modelagem, treinamento e teste.

### Identificar atributos e variáveis-alvo
É comum que os dados que você obteve contenham atributos não numéricos. Isso pode ser um problema, dado que a maioria dos algoritmos de machine learning esperam dados númericos para operar cálculos.

Execute a célula de código abaixo para separar os dados dos estudantes em atributos e variáveis-alvo e verificar se algum desses atributos é não numérico.

In [51]:
# Extraia as colunas dos atributo
feature_cols = list(student_data.columns[:-1])

# Extraia a coluna-alvo 'passed'
target_col = student_data.columns[-1] 

# Mostre a lista de colunas
print ("Colunas de atributos:\n{}".format(feature_cols))
print ("\nColuna-alvo: {}".format(target_col))

# Separe os dados em atributos e variáveis-alvo (X_all e y_all, respectivamente)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Mostre os atributos imprimindo as cinco primeiras linhas
print ("\nFeature values:")
X_all.head()

Colunas de atributos:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Coluna-alvo: passed

Feature values:


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       yes      yes        no       5         3     3    1    1      3   
2   ...       yes      yes        no       4         3     2    2    3      3   
3   ...       yes      yes       yes       3         2     2    1    1      5   
4   ...       yes       no        no       4         3     2    1    2      5   

  absences  
0        6  
1        4  
2       10  
3        2  
4        4  

[5 rows x 30 columns]

### Pré-processar Colunas de Atributo

Como você pode ver, há muitas colunas não numéricas que precisam ser convertidas! Muitas delas são simplesmente `yes`/`no`, por exemplo, a coluna `internet`. É razoável converter essas variáveis em valores (binários) `1`/`0`.

Outras colunas, como `Mjob` e `Fjob`, têm mais do que dois valores e são conhecidas como variáveis categóricas. A maneira recomendada de lidar com esse tipo de coluna é criar uma quantidade de colunas proporcional aos possíveis valores (por exemplo, `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc), e assinalar `1` para um deles e `0` para todos os outros.

Essas colunas geradas são por vezes chamadas de _variáveis postiças_ (_dummy variables_), e nós iremos utilizar a função [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) para fazer essa conversão. Execute a célula de código abaixo para executar a rotina de pré-processamento discutida nesta seção.

In [52]:
def preprocess_features(X):
    ''' Pré-processa os dados dos estudantes e converte as variáveis binárias não numéricas em
        variáveis binárias (0/1). Converte variáveis categóricas em variáveis postiças. '''
    
    # Inicialize nova saída DataFrame
    output = pd.DataFrame(index = X.index)

    # Observe os dados em cada coluna de atributos 
    for col, col_data in X.iteritems():
        
        # Se o tipo de dado for não numérico, substitua todos os valores yes/no por 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # Se o tipo de dado for categórico, converta-o para uma variável dummy
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Reúna as colunas revisadas
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print ("Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns)))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


In [53]:
X_all.head()

school_GP  school_MS  sex_F  sex_M  age  address_R  address_U  famsize_GT3  \
0          1          0      1      0   18          0          1            1   
1          1          0      1      0   17          0          1            1   
2          1          0      1      0   15          0          1            0   
3          1          0      1      0   15          0          1            1   
4          1          0      1      0   16          0          1            1   

   famsize_LE3  Pstatus_A    ...     higher  internet  romantic  famrel  \
0            0          1    ...          1         0         0       4   
1            0          0    ...          1         1         0       5   
2            1          0    ...          1         1         0       4   
3            0          0    ...          1         1         1       3   
4            0          0    ...          1         0         0       4   

   freetime  goout  Dalc  Walc  health  absences  
0         3      4     1     1       3         6  
1         3      3     1     1       3         4  
2         3      2     2     3       3        10  
3         2      2     1     1       5         2  
4         3      2     1     2       5         4  

[5 rows x 48 columns]

In [54]:
X_all.tail()

school_GP  school_MS  sex_F  sex_M  age  address_R  address_U  \
390          0          1      0      1   20          0          1   
391          0          1      0      1   17          0          1   
392          0          1      0      1   21          1          0   
393          0          1      0      1   18          1          0   
394          0          1      0      1   19          0          1   

     famsize_GT3  famsize_LE3  Pstatus_A    ...     higher  internet  \
390            0            1          1    ...          1         0   
391            0            1          0    ...          1         1   
392            1            0          0    ...          1         0   
393            0            1          0    ...          1         1   
394            0            1          0    ...          1         1   

     romantic  famrel  freetime  goout  Dalc  Walc  health  absences  
390         0       5         5      4     4     5       4        11  
391         0       2         4      5     3     4       2         3  
392         0       5         5      3     3     3       3         3  
393         0       4         4      1     3     4       5         0  
394         0       3         2      3     3     3       5         5  

[5 rows x 48 columns]

### Implementação: Divisão dos Dados de Treinamento e Teste
Até agora, nós convertemos todos os atributos _categóricos_ em valores numéricos. Para o próximo passo, vamos dividir os dados (tanto atributos como os rótulos correspondentes) em conjuntos de treinamento e teste. Na célula de código abaixo, você irá precisar implementar o seguinte:
- Embaralhe aleatoriamente os dados (`X_all`, `y_all`) em subconjuntos de treinamento e teste.
  - Utilizar 300 pontos de treinamento (aproxidamente 75%) e 95 pontos de teste (aproximadamente 25%).
  - Estabelecer um `random_state` para as funções que você utiliza, se a opção existir.
  - Armazene os resultados em `X_train`, `X_test`, `y_train` e `y_test`.

In [55]:
# TODO: Importe qualquer funcionalidade adicional de que você possa precisar aqui
from sklearn.cross_validation import train_test_split
# TODO: Estabeleça o número de pontos de treinamento
num_train = 300

# Estabeleça o número de pontos de teste
num_test = X_all.shape[0] - num_train
print(num_test)

# TODO: Emabaralhe e distribua o conjunto de dados de acordo com o número de pontos de treinamento e teste abaixo
X_train, X_test, y_train, y_test = train_test_split(
  X_all, y_all, test_size=0.24, train_size = 0.76, random_state=42)

# Mostre o resultado da distribuição
print ("O conjunto de treinamento tem {} amostras.".format(X_train.shape[0]))
print ("O conjunto de teste tem {} amostras.".format(X_test.shape[0]))

95
O conjunto de treinamento tem 300 amostras.
O conjunto de teste tem 95 amostras.


## Treinando e Avaliando Modelos
Nesta seção, você irá escolher 3 modelos de aprendizagem supervisionada que sejam apropriados para esse problema e que estejam disponíveis no `scikit-learn`. Primeiro você irá discutir o raciocínio por trás da escolha desses três modelos considerando suas vantagens e desvantagens e o que você sabe sobre os dados. Depois você irá ajustar o modelo a diferentes tamanhos de conjuntos de treinamento (com 100, 200 e 300 pontos) e medir a pontuação F<sub>1</sub>. Você vai precisar preencher três tabelas (uma para cada modelo) que mostrem o tamanho do conjunto de treinamento, o tempo de treinamento, o tempo de previsão e a pontuação F<sub>1</sub> no conjunto de treinamento.

**Os seguintes modelos de aprendizagem supervisionada estão atualmente disponíveis no **[`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html)** para você escolher:**
- Gaussian Naive Bayes (GaussianNB)
- Árvores de Decisão
- Métodos de agregação (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Método do gradiente estocástico (SGDC)
- Máquinas de vetores de suporte (SVM)
- Regressão logística

### Questão 2 - Aplicação dos Modelos
*Liste três modelos de aprendizagem supervisionada que são apropriadas para esse problema. Para cada modelo escolhido:*
- Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. *(Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)* 
- Quais são as vantagens do modelo; quando ele tem desempenho melhor? 
- Quais são as desvantagens do modelo, quando ele tem desempenho pior?
- O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?

**Resposta: 
Árvore de Decisão: 

Um exemplo de aplicação encontrado foi o uso do modelo na produção de laranjas. O modelo possibilitou encontrar alternativas de vendas economicamente mais viáveis (http://www.custoseagronegocioonline.com.br/numero4v9/decisao.pdf).

Vantagens: Facilidade de compreensão e interpretação, lida com vários tipos de variáveis, destaca os atributos mais relevantes, adaptável a problemas de regressão;

Desvantagens: sensibilidade a pequenas perturbações no conjunto de treino, não fornece um quadro nítido da estrutura,  limitação a uma linguagem descritiva baseada em atributos-valores, um mesmo conceito pode ser representado por várias árvores de decisão, modelo propenso ao overfitting;

Como no conjunto de dados forncecidos existem variáveis de vários tipos que podem influenciar a aprovação ou reprovação do aluno, esse modelo pode selecionar os atributos mais importantes na classificação dos alunos em risco de reprovação e possibilitar a tomada de ação.


SVM: 
Aplicação: predição do risco de morte de pacientes internados com síndrome coronariana aguda (http://www.din.uem.br/sbpo/sbpo2010/pdf/72295.pdf).

SVMs efetua transformações múltiplas nos dados a fim de encontrar uma divisão baseada nos rótulos dos dados. Esse modelo busca encontrar um hiperplano que separa linearmente de maneira ótima o conjunto de dados. Além disso, os dados que não saão linearmente separáveis são transformados através de truques de Kernel, para que uma separação seja alcançada.

SVMs não são afetados por mínimos locais e se adaptam bem a muitas dimensões.

Vetores de suporte, os pontos mais próximos à região de decisão são difíceis de classificar.

SVM foi uma opção devido à quantidade de atributos envolvidos na classificação.


Logistic Regression:

Aplicação: análise de mercado em que a variável resposta é dicotómica entre consumir, ou não, produtos de determinada marca (http://repositorio.ul.pt/bitstream/10451/10671/1/ulfc106455_tm_Cleidy_Cabral.pdf). 

O modelo de regressão logística é semelhante ao modelo de regressão linear. No entanto, no modelo logístico a variável resposta é binária.

Vantagens: análise logística controla grande número de variáveis simultaneamente, permitindo que os dados sejam utilizados mais eficientemente; a discriminação logística é mais genérica; número reduzido de parâmetros;

Desvantagem: O modelo logístico é bastante sensível à colinearidade.

A regressão logística foi considerada pelo fato da variável resposta ser binária e existirem muitos parâmetros envolvidos.
**

### Configuração
Execute a célula de código abaixo para inicializar três funções de ajuda que você pode utilizar para treinar e testar os três modelos de aprendizagem supervisionada que você escolheu acima. As funções são as seguintes:
- `train_classifier` - recebe como parâmetro um classificador e dados de treinamento e ajusta o classificador aos dados.
- `predict_labels` - recebe como parâmetro um classificador ajustado, atributos e rótulo alvo e faz estimativas utilizando a pontuação do F<sub>1</sub>.
- `train_predict` - recebe como entrada um classificador, e dados de treinamento e teste, e executa `train_clasifier` e `predict_labels`.
 - Essa função vai dar a pontuação F<sub>1</sub> tanto para os dados de treinamento como para os de teste, separadamente.

In [56]:
import time
def train_classifier(clf, X_train, y_train):
    ''' Ajusta um classificador para os dados de treinamento. '''
    
    print ("Treinando {}...".format(clf.__class__.__name__))
    # Inicia o relógio, treina o classificador e, então, para o relógio
    start = time.time()
    clf.fit(X_train, y_train)
    end = time.time()
    print ("Tempo de treinamento: (secs): {:.3f}".format(end - start))
    
    # Imprime os resultados
    print ("O modelo foi treinado em {:.4f} segundos".format(end - start))

    
def predict_labels(clf, features, target):
    ''' Faz uma estimativa utilizando um classificador ajustado baseado na pontuação F1. '''
    
    # Inicia o relógio, faz estimativas e, então, o relógio para
    start = time.time()
    y_pred = clf.predict(features)
    end = time.time()
    
    # Imprime os resultados de retorno
    print ("As previsões foram feitas em {:.4f} segundos.".format(end - start))
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Treina e faz estimativas utilizando um classificador baseado na pontuação do F1. '''
    
    # Indica o tamanho do classificador e do conjunto de treinamento
    print ("Treinando um {} com {} pontos de treinamento. . .".format(clf.__class__.__name__, len(X_train)))
    
    # Treina o classificador
    train_classifier(clf, X_train, y_train)
    
    # Imprime os resultados das estimativas de ambos treinamento e teste
    print ("Pontuação F1 para o conjunto de treino: {:.4f}.".format(predict_labels(clf, X_train, y_train)))
    print ("Pontuação F1 para o conjunto de teste: {:.4f}.".format(predict_labels(clf, X_test, y_test)))

### Implementação: Métricas de Desempenho do Modelo
Com as funções acima, você vai importar os três modelos de aprendizagem supervisionada de sua escolha e executar a função `train_prediction` para cada um deles. Lembre-se de que você vai precisar treinar e usar cada classificador para três diferentes tamanhos de conjuntos de treinamentos: 100, 200 e 300 pontos. Então você deve ter 9 saídas diferentes abaixo – 3 para cada modelo utilizando cada tamanho de conjunto de treinamento. Na célula de código a seguir, você deve implementar o seguinte:
- Importe os três modelos de aprendizagem supervisionada que você escolheu na seção anterior.
- Inicialize os três modelos e armazene eles em `clf_A`, `clf_B` e `clf_C`.
 - Defina um `random_state` para cada modelo, se a opção existir.
 - **Nota:** Utilize as configurações padrão para cada modelo – você vai calibrar um modelo específico em uma seção posterior.
- Crie diferentes tamanhos de conjuntos de treinamento para treinar cada modelo.
 - *Não embaralhe e distribua novamente os dados! Os novos pontos de treinamento devem ser tirados de `X_train` e `y_train`.*
- Treine cada modelo com cada tamanho de conjunto de treinamento e faça estimativas com o conjunto de teste (9 vezes no total).  
**Nota:** Três tabelas são fornecidas depois da célula de código a seguir, nas quais você deve anotar seus resultados.

In [57]:
# TODO: Importe os três modelos de aprendizagem supervisionada do sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.linear_model import LogisticRegression

# TODO: Inicialize os três modelos
clf_A = DecisionTreeClassifier()
clf_B = svm.SVC()
clf_C = LogisticRegression()

# TODO: Configure os tamanho dos conjuntos de treinamento
X_train_100, X_test_100, y_train_100, y_test_100 = train_test_split(
  X_all, y_all, test_size=0.745, train_size = 0.255, random_state=42)

# Mostre o resultado da distribuição
print("")
print ("O conjunto de treinamento 1 tem {} amostras.".format(X_train_100.shape[0]))
print ("O conjunto de teste 1 tem {} amostras.".format(X_test_100.shape[0]))

X_train_200, X_test_200, y_train_200, y_test_200 = train_test_split(
  X_all, y_all, test_size=0.493, train_size = 0.507, random_state=42)

# Mostre o resultado da distribuição
print("")
print ("O conjunto de treinamento 2 tem {} amostras.".format(X_train_200.shape[0]))
print ("O conjunto de teste 2 tem {} amostras.".format(X_test_200.shape[0]))

X_train_300, X_test_300, y_train_300, y_test_300 = train_test_split(
  X_all, y_all, test_size=0.24, train_size = 0.76, random_state=42)

# Mostre o resultado da distribuição
print("")
print ("O conjunto de treinamento 2 tem {} amostras.".format(X_train_300.shape[0]))
print ("O conjunto de teste 2 tem {} amostras.".format(X_test_300.shape[0]))

# TODO: Executar a função 'train_predict' para cada classificador e cada tamanho de conjunto de treinamento
# train_predict(clf, X_train, y_train, X_test, y_test)



O conjunto de treinamento 1 tem 100 amostras.
O conjunto de teste 1 tem 295 amostras.

O conjunto de treinamento 2 tem 200 amostras.
O conjunto de teste 2 tem 195 amostras.

O conjunto de treinamento 2 tem 300 amostras.
O conjunto de teste 2 tem 95 amostras.


In [58]:
print("")
train_predict(clf_A, X_train_100, y_train_100, X_test_100, y_test_100)
print("")
train_predict(clf_A, X_train_200, y_train_200, X_test_200, y_test_200)
print("")
train_predict(clf_A, X_train_300, y_train_300, X_test_300, y_test_300)


Treinando um DecisionTreeClassifier com 100 pontos de treinamento. . .
Treinando DecisionTreeClassifier...
Tempo de treinamento: (secs): 0.002
O modelo foi treinado em 0.0020 segundos
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de treino: 1.0000.
As previsões foram feitas em 0.0000 segundos.
Pontuação F1 para o conjunto de teste: 0.6957.

Treinando um DecisionTreeClassifier com 200 pontos de treinamento. . .
Treinando DecisionTreeClassifier...
Tempo de treinamento: (secs): 0.010
O modelo foi treinado em 0.0100 segundos
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de treino: 1.0000.
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de teste: 0.7704.

Treinando um DecisionTreeClassifier com 300 pontos de treinamento. . .
Treinando DecisionTreeClassifier...
Tempo de treinamento: (secs): 0.013
O modelo foi treinado em 0.0130 segundos
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conju

In [59]:
print("")
train_predict(clf_B, X_train_100, y_train_100, X_test_100, y_test_100)
print("")
train_predict(clf_B, X_train_200, y_train_200, X_test_200, y_test_200)
print("")
train_predict(clf_B, X_train_300, y_train_300, X_test_300, y_test_300)


Treinando um SVC com 100 pontos de treinamento. . .
Treinando SVC...
Tempo de treinamento: (secs): 0.004
O modelo foi treinado em 0.0040 segundos
As previsões foram feitas em 0.0040 segundos.
Pontuação F1 para o conjunto de treino: 0.8481.
As previsões foram feitas em 0.0070 segundos.
Pontuação F1 para o conjunto de teste: 0.8115.

Treinando um SVC com 200 pontos de treinamento. . .
Treinando SVC...
Tempo de treinamento: (secs): 0.010
O modelo foi treinado em 0.0100 segundos
As previsões foram feitas em 0.0090 segundos.
Pontuação F1 para o conjunto de treino: 0.8868.
As previsões foram feitas em 0.0060 segundos.
Pontuação F1 para o conjunto de teste: 0.8013.

Treinando um SVC com 300 pontos de treinamento. . .
Treinando SVC...
Tempo de treinamento: (secs): 0.026
O modelo foi treinado em 0.0260 segundos
As previsões foram feitas em 0.0170 segundos.
Pontuação F1 para o conjunto de treino: 0.8761.
As previsões foram feitas em 0.0040 segundos.
Pontuação F1 para o conjunto de teste: 0.7838

In [60]:
print("")
train_predict(clf_C, X_train_100, y_train_100, X_test_100, y_test_100)
print("")
train_predict(clf_C, X_train_200, y_train_200, X_test_200, y_test_200)
print("")
train_predict(clf_C, X_train_300, y_train_300, X_test_300, y_test_300)


Treinando um LogisticRegression com 100 pontos de treinamento. . .
Treinando LogisticRegression...
Tempo de treinamento: (secs): 0.004
O modelo foi treinado em 0.0040 segundos
As previsões foram feitas em 0.0030 segundos.
Pontuação F1 para o conjunto de treino: 0.9275.
As previsões foram feitas em 0.0100 segundos.
Pontuação F1 para o conjunto de teste: 0.7196.

Treinando um LogisticRegression com 200 pontos de treinamento. . .
Treinando LogisticRegression...
Tempo de treinamento: (secs): 0.006
O modelo foi treinado em 0.0060 segundos
As previsões foram feitas em 0.0520 segundos.
Pontuação F1 para o conjunto de treino: 0.8721.
As previsões foram feitas em 0.0150 segundos.
Pontuação F1 para o conjunto de teste: 0.7814.

Treinando um LogisticRegression com 300 pontos de treinamento. . .
Treinando LogisticRegression...
Tempo de treinamento: (secs): 0.038
O modelo foi treinado em 0.0380 segundos
As previsões foram feitas em 0.0530 segundos.
Pontuação F1 para o conjunto de treino: 0.8468.
A

### Resultados em tabelas
Edite a célula abaixo e veja como a tabela pode ser desenhada em [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). Você deve salvar seus resultados abaixo nas tabelas fornecidas.

**DecisionTreeClassifier**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento (segundos)| Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100 |0.0030|0.0010|1.0000|0.6950|
| 200 |0.004|0.0010|1.0000|0.7528|
| 300 |0.0140|0.0010|1.0000|0.5983|

** SVM **  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento (segundos)| Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100 |0.0050|0.0040|0.8481|0.8115|
| 200 |0.0270|0.0050|0.8868|0.8013|
| 300 |0.0600|0.0040|0.8761|0.7838|

** LogisticRegression **  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento (segundos)| Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100 |0.0030|0.0020|0.9275|0.7196|
| 200 |0.0060|0.0010|0.8721|0.7814|
| 300 |0.0220|0.0020|0.8468|0.8060|

## Escolhendo o Melhor Modelo
Nesta seção final, você irá escolher dos três modelos de aprendizagem supervisionada o *melhor* para utilizar os dados dos estudantes. Você então executará um busca em matriz otimizada para o modelo em todo o conjunto de treinamento (`X_train` e `y_train`) ao calibrar pelo menos um parâmetro, melhorando em comparação a pontuação F<sub>1</sub> do modelo não calibrado. 

### Questão 3 - Escolhendo o Melhor Modelo
*Baseando-se nos experimentos que você executou até agora, explique em um ou dois parágrafos ao conselho de supervisores qual modelo que você escolheu como o melhor. Qual modelo é o mais apropriado baseado nos dados disponíveis, recursos limitados, custo e desempenho?*

**Resposta: Baseado no conjunto de dados obtidos até agora, considerando que, em geral, o modelo SVM resultou em maiores pontuações de teste, essa seria a minha indicação para os supervisores. Esse modelo "funciona ao mapear dados para um espaço de variável altamente dimensional para que os pontos de dados possam ser categorizados, mesmo quando os dados não forem de outra forma linearmente separáveis"(IBM). O custo é relativamente baixo, uma vez que foram necessários apenas 0.004 segundos para testar o modelo aplicado a 95 alunos e 0.06 segundos para treinar um modelo com 300 alunos. Seriam necessários apenas 6.73 e 24 segundos, aproximadamente para testar e treinar respectivamente o modelo em uma escola com 1600 alunos, por exemplo.**

### Questão 4 – O Modelo para um Leigo
*Em um ou dois parágrafos, explique para o conselho de supervisores, utilizando termos leigos, como o modelo final escolhido deve trabalhar. Tenha certeza que você esteja descrevendo as melhores qualidades do modelo, por exemplo, como o modelo é treinado e como ele faz uma estimativa. Evite jargões técnicos ou matemáticos, como descrever equações ou discutir a implementação do algoritmo.*

**Resposta: Support Vector Machines (SVM), é um modelo que classifica membros de um conjunto de dados em diferentes classes de associação. 
![alt text](img1.png)
Na imagem acima, os objetos pertencem à classe verde ou vermeho. A linha que separa os conjuntos define uma fronteira de separação, de forma que qualquer novo objeto que caia no lado direito da região será classificado como verde. Esse é um exemplo simples em que o conjunto é linearmente separável. No entanto, muitos problemas de classificação não são linearmente separáveis e exigem estruturas mais complexas para uma separação ótima, como mostra a imagem abaixo:
![alt text](img2.png)
Problemas como esse exigem uma separação em hiperplanos e são facilmente tratados através de modelos SVM. Esse modelo utiliza funções matemáticas, conhecidas como "truques de kernel", para rearranjar os objetos, de forma que eles sejam linearmente separáveis. A figura abaixo mostra como o truque de Kernel funciona. Ao invés de construir um modelo complexo que se adapte às curvas de separação do problema original (figura à esquerda), com o truque de Kernel basta encontrar uma linha ótima que separe os objetos verdes e vermelhos.
![alt text](img3.png)
fonte: http://www.statsoft.com/Textbook/Support-Vector-Machines
**

### Implementação: Calibrando o Modelo (_Tuning_)
Calibre o modelo escolhido. Utilize busca em matriz (`GridSearchCV`) com, pelo menos, um parâmetro importante calibrado com, pelo menos, 3 valores diferentes. Você vai precisar utilizar todo o conjunto de treinamento para isso. Na célula de código abaixo, você deve implementar o seguinte:
- Importe [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) e [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Crie o dicionário de parâmetros que você deseja calibrar para o modelo escolhido.
 - Examplo: `parameters = {'parameter' : [list of values]}`.
- Inicialize o classificador que você escolheu e armazene-o em `clf`.
- Crie a função de pontuação F<sub>1</sub> utilizando `make_scorer` e armazene-o em `f1_scorer`.
 - Estabeleça o parâmetro `pos_label` para o valor correto!
- Execute uma busca em matriz no classificador `clf` utilizando o `f1_scorer` como método de pontuação e armazene-o em `grid_obj`.
- Treine o objeto de busca em matriz com os dados de treinamento (`X_train`, `y_train`) e armazene-o em `grid_obj`.

In [61]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {'C' : [2.0, 3.0,4.0, 10.0,50.0], 'kernel': ['poly', 'rbf', 'sigmoid']}

# TODO: Initialize the classifier
clf = svm.SVC(random_state=42)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring=f1_scorer, cv=5, verbose=0)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Reporte a pontuação final F1 para treinamento e teste depois de calibrar os parâmetrosprint "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print ("O modelo calibrado tem F1 de {:.4f} no conjunto de treinamento.".format(predict_labels(clf, X_train, y_train)))
print ("O modelo calibrado tem F1 de {:.4f} no conjunto de teste.".format(predict_labels(clf, X_test, y_test)))

As previsões foram feitas em 0.0100 segundos.
O modelo calibrado tem F1 de 0.9318 no conjunto de treinamento.
As previsões foram feitas em 0.0040 segundos.
O modelo calibrado tem F1 de 0.7737 no conjunto de teste.


### Questão 5 - Pontuação F<sub>1</sub> Final
*Qual é a pontuação F<sub>1</sub> do modelo final para treinamento e teste? Como ele se compara ao modelo que não foi calibrado?*

**Resposta: O modelo final tem pontuação F1 0.9318 de treinamento e 0.7737 de teste. A pontuação de teste foi ligeiramente menor em comparação ao modelo não calibrado.**

> **Nota**: Uma vez que você completou todas as implementações de código e respondeu todas as questões acima com êxito, você pode finalizar seu trabalho exportando o iPython Nothebook como um document HTML. Você pode fazer isso utilizando o menu acima e navegando para  
**File -> Download as -> HTML (.html)**. Inclua a documentação final junto com o notebook para o envio do seu projeto.